### PR TITLE
feat(frontend): `StreamDynamicFilter`

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

- Summarize your change: Add `StreamDynamicFilter`, converting from a non-eq inner join with comparison operator
- Changes `to_stream_plan`

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
Tracking issue: https://github.com/singularity-data/risingwave/issues/3419

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 1]
```diff
@@ -772,10 +774,79 @@ impl ToStream for LogicalJoin {
...
+
+            let left = self
+                .left()
+                .to_stream_with_dist_required(&RequiredDist::shard_by_key(
+                    self.left().schema().len(),
+                    &[left_ref_index],
+                ))?;
+
+            // The right input needs to be a direct descendant of a simple agg
+            assert!(*self.right().distribution() == Distribution::Single);
```
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:828]:
```suggestion
```
I think here the right input is just a logical plan node which has no distribution property at all.

Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:827]: Hmm, I think every plan node has a distribution, though? 

I think this assert could catch some errors. We can't expect single value to have distribution other than single.

File: src/frontend/src/optimizer/plan_node/logical_join.rs [hunk 2]
```diff
@@ -772,10 +774,78 @@ impl ToStream for LogicalJoin {
...
+
+            let right = self
+                .right()
+                .to_stream_with_dist_required(&RequiredDist::PhysicalDist(
+                    Distribution::Broadcast,
+                ))?;
+
+            let plan = StreamDynamicFilter::new(predicate.other_cond().clone(), left, right).into();
+
+            if self.output_indices != (0..self.internal_column_num()).collect::<Vec<_>>() {
```
Comment (jon-chuang) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:837]: Actually, we should probably use the rewritten logical join and output indices

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:827]: Oh I see what you are saying. Maybe I should check that the right side after converted to stream has input that is distribution single...

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:827]: > every plan node has a distribution

Probably not. the distribution for a logical node is meaningless and a logical node will not give any guarantee about its distribution.

File: src/frontend/test_runner/tests/testdata/tpch.yaml [hunk 3]
```diff
@@ -1008,6 +1008,42 @@
...
+                          StreamExchange { dist: HashShard([1]) }
+                            StreamTableScan { table: partsupp, columns: [ps_partkey, ps_suppkey, ps_availqty, ps_supplycost, _row_id], pk_indices: [4] }
+                          StreamExchange { dist: HashShard([0]) }
+                            StreamTableScan { table: supplier, columns: [s_suppkey, s_nationkey, _row_id], pk_indices: [2] }
+                      StreamExchange { dist: HashShard([0]) }
+                        StreamProject { exprs: [$0, $1] }
+                          StreamFilter { predicate: ($2 = 'ARGENTINA':Varchar) }
+                            StreamTableScan { table: nation, columns: [n_nationkey, _row_id, n_name], pk_indices: [1] }
+            StreamProject { exprs: [($1 * 0.0001000000:Decimal)] }
+              StreamExchange { dist: Broadcast }
```
Comment (jon-chuang) [inline src/frontend/test_runner/tests/testdata/tpch.yaml:1032]: For some weird reason, `Exchange` is the chiild of `Project`. Trying to figure out why...

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/tpch.yaml:1032]: It's weird, this scenario only occurs for `Broadcast` but not, for instance `SomeShard`...

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/tpch.yaml:1032]: Any ideas? @st1page

Comment (jon-chuang) [reply src/frontend/test_runner/tests/testdata/tpch.yaml:1032]: Ok, fixed, turns out it was not matching on Broadcast somewhere...

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 2 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:837]: actually, they are identical, so this is unnecessary. simplified some existing code

File: src/frontend/src/optimizer/plan_node/logical_join.rs (hunk 1 above)
Comment (jon-chuang) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:827]: Ok, I've moved the check to after the conversion of right side to stream plan

**Comment by jon-chuang** [comment]
Hey @st1page are there any issues with this? Next PR will be able to pass e2e tests so if not could we merge this one?

**Review by st1page — APPROVED** [review]
LGTM

**Pull request merged**